### PR TITLE
Refactor compliance connection to sensor

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

In the current state, compliance calls the `Communicate` function every time it wants to send something. This will orphan the gRPC stub created in `manageReceivedStream`, thus making sensor unable to send any messages to compliance.

This PR aims to solve this problem by reusing the same gRPC stub for receiving and sending messages from/to sensor (tracked in [ROX-15100](https://issues.redhat.com/browse/ROX-15100)).

See: 
* [ROX-13383](https://issues.redhat.com/browse/ROX-13383)

## Checklist
- [ ] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

* Manual testing in a maple team mob session
    * Build ACS with the following variables:
     ```
     ROX_POSTGRES_DATASTORE         = true
     ROX_RHCOS_FAKE_NODE_INVENTORY  = true
     ROX_NODE_RESCAN_INTERVAL       = 10s
     ROX_RHCOS_NODE_SCANNING        = true
     ```
    * Deploy with ACS with `deploy/k8s/deploy-local.sh`
    * Set the re-scan interval with: `kubectl -n stackrox set env ds/collector ROX_NODE_RESCAN_INTERVAL=$ROX_NODE_RESCAN_INTERVAL`
    * We should see logs in sensor and compliance of the scans being done.
    * In the UI go to `Compliance` and click on `Scan Environment`
    * We should see logs in sensor and compliance of the `scrape` being started.
    * Kill sensor and check everything should recover.
    * Kill compliance and check everything should recover.

### CODE REVIEW

File: compliance/collection/main.go [hunk 1]
```diff
@@ -109,32 +109,16 @@ func startAuditLogCollection(ctx context.Context, client sensor.ComplianceServic
...
-	}
-}
-
-func manageReceiveStream(ctx context.Context, cli sensor.ComplianceServiceClient, sig *concurrency.Signal) {
+func manageStream(ctx context.Context, cli sensor.ComplianceServiceClient, sig *concurrency.Signal, sensorC <-chan *sensor.MsgFromCompliance) {
 	for {
 		select {
 		case <-ctx.Done():
 			sig.Signal()
+			log.Infof("Parent context cancelled. Stopping manageStream")
```
Comment (vikin91) [inline compliance/collection/main.go:117]: I think we may remove this log line now. It was useful for understanding the behavior in the debug session, but let's not bother our users with it.

File: compliance/collection/main.go [hunk 2]
```diff
@@ -143,9 +127,31 @@ func manageReceiveStream(ctx context.Context, cli sensor.ComplianceServiceClient
...
+			cancelFn() // runRecv is blocking, so the context is safely cancelled before the next  call to initializeStream
+		}
+	}
+}
+
+func manageSendToSensor(ctx context.Context, cli sensor.ComplianceService_CommunicateClient, sensorC <-chan *sensor.MsgFromCompliance) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("Child context cancelled. Stopping manageSendToSensor")
```
Comment (vikin91) [inline compliance/collection/main.go:149]:
```suggestion
```

File: compliance/collection/main.go [hunk 3]
```diff
@@ -143,9 +127,31 @@ func manageReceiveStream(ctx context.Context, cli sensor.ComplianceServiceClient
...
+			return
+		case sc := <-sensorC:
+			if err := cli.Send(sc); err != nil {
+				log.Errorf("failed sending nodeScanV2 to sensor: %v", err)
```
Comment (vikin91) [inline compliance/collection/main.go:153]: change `nodeScanV2` to either `node scan` or `node inventory`

File: compliance/collection/main.go [hunk 4]
```diff
@@ -278,15 +284,13 @@ func main() {
...
-	go manageReceiveStream(ctx, cli, &stoppedSig)
-
 	if features.RHCOSNodeScanning.Enabled() {
 		rescanInterval := env.NodeRescanInterval.DurationSetting()
 		cmetrics.ObserveRescanInterval(rescanInterval, getNode())
 		log.Infof("Node Rescan interval: %s", rescanInterval.String())
 		sensorC := make(chan *sensor.MsgFromCompliance)
 		defer close(sensorC)
-		go manageSendingToSensor(ctx, cli, sensorC)
+		go manageStream(ctx, cli, &stoppedSig, sensorC)
```
Comment (vikin91) [inline compliance/collection/main.go:293]: We could move those 3 lines before the `if features.RHCOSNodeScanning.Enabled()` and remove the `else` clause in line 315.

**Review by vikin91 — COMMENTED** [review]
Sharing quickly my initial thoughts. Will do a thorough review later.

**Review by vikin91 — APPROVED** [review]
The code looks good now! Could we give it a test on a local cluster before we merge?

**Comment by Maddosaurus** [comment]
Yes, I tested the changes as @lvalerom described.
Deployed to Colima, tested killing the Sensor as well as Compliance and observed all connections recovering as well as the UI scan functionality still working.

**Pull request merged**